File: lib/contrek/reducers/uniq_reducer.rb
module Contrek
  module Reducers
    class UniqReducer < Reducer
      def reduce!
        @points.replace @points.reject.with_index { |e, i| i < (@points.size - 1) and e[:x] == @points[i + 1][:x] && e[:y] == @points[i + 1][:y] }
      end
    end
  end
end

File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/reducers/UniqReducer.cpp
/*
 * UniqReducer.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include <vector>
#include <algorithm>
#include <iostream>
#include "UniqReducer.h"
#include "Reducer.h"

UniqReducer::UniqReducer(std::vector<Point*>& list_of_points) : Reducer(list_of_points) {
}

struct is_near {
  bool operator() (Point* first, Point* second) const {
    return first->x == second->x && first->y == second->y;
  }
};

void UniqReducer::reduce() {
  auto last = std::unique(points.begin(), points.end(), is_near());
  points.erase(last, points.end());
}
